File: bubbler_app/lib/data/repositories/feed_repository.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_exception.dart';
import '../../core/api/endpoints.dart';
import '../models/graph.dart';
import '../models/post.dart';

/// Ranked feed and graph session seed (`getFeed` / `getSessionFeed`).
class FeedRepository {
  FeedRepository(this._client);

  final ApiClient _client;

  /// Ranked feed. Optional [query] maps to `q` and seeds similar/opposite
  /// candidates.
  Future<List<Post>> getFeed({String? query}) async {
    final trimmed = query?.trim() ?? '';
    final data = await _client.authorizedRequest(
      path: Endpoints.feedMe,
      queryParameters: trimmed.isEmpty ? null : {'q': trimmed},
    );
    return _decodePosts(data);
  }

  /// Graph session seed queue (`GET /feed/me/session`).
  Future<GraphSessionFeed> getSessionFeed({bool diversify = false}) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.feedMeSession,
      queryParameters: diversify ? {'diversify': 'true'} : null,
    );
    return GraphSessionFeed.fromJson(_asJsonMap(data));
  }

  static List<Post> _decodePosts(dynamic data) {
    if (data is! List) {
      throw const ApiInvalidResponse();
    }
    return data
        .map((e) => Post.fromJson(_asJsonMap(e)))
        .toList();
  }

  static Map<String, dynamic> _asJsonMap(dynamic data) {
    if (data is Map<String, dynamic>) {
      return data;
    }
    if (data is Map) {
      return Map<String, dynamic>.from(data);
    }
    throw const ApiInvalidResponse();
  }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/repository/FeedRepository.kt
package com.bubbler.android.data.repository

import com.bubbler.android.core.auth.TokenStore
import com.bubbler.android.core.network.ApiClient
import com.bubbler.android.core.network.Endpoints
import com.bubbler.android.data.model.GraphSessionFeed
import com.bubbler.android.data.model.Post

/** Ranked feed + graph session seed (`getFeed`, `getSessionFeed`). */
open class FeedRepository(
    private val apiClient: ApiClient,
    private val tokenStore: TokenStore,
) {
    /** Ranked feed. Optional [query] maps to `q` and seeds similar/opposite candidates. */
    open suspend fun getFeed(query: String? = null): List<Post> {
        val trimmed = query?.trim().orEmpty()
        val path = Endpoints.withQuery(
            Endpoints.FEED_ME,
            "q" to trimmed.ifEmpty { null },
        )
        return apiClient.get(path, token = tokenStore.requireAccessToken())
    }

    open suspend fun getSessionFeed(diversify: Boolean = false): GraphSessionFeed {
        val path = Endpoints.withQuery(
            Endpoints.FEED_ME_SESSION,
            "diversify" to if (diversify) "true" else null,
        )
        return apiClient.get(path, token = tokenStore.requireAccessToken())
    }
}
